File: crates/yokan/demo/calcgrid.py
# /// script
# requires-python = ">=3.14"
# ///
"""The calculator again, on a grid. `grid(columns=4, rows=5)`
lays equal tracks, every key fills its cell, and the zero key says
`col_span=2` — the whole keypad is one container instead of five
rows. Same store as demo/calc.py; only the view differs."""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import button, column, grid, run, store, strings, style, text  # noqa: E402


@store
class Calc:
    display: str = "0"
    acc: float = 0.0
    op: str = ""
    fresh: bool = True
    has_dot: bool = False

    def press(self, d: str) -> None:
        if self.fresh:
            self.display = d
            self.fresh = False
            self.has_dot = False
        elif self.display == "0":
            self.display = d
        else:
            self.display = self.display + d

    def dot(self) -> None:
        if self.fresh:
            self.display = "0."
            self.fresh = False
            self.has_dot = True
        elif not self.has_dot:
            self.display = self.display + "."
            self.has_dot = True

    def negate(self) -> None:
        v = strings.to_float(self.display, 0.0)
        if v != 0.0:
            self.display = f"{0.0 - v}"
            self.fresh = False

    def percent(self) -> None:
        v = strings.to_float(self.display, 0.0)
        self.display = f"{v / 100.0}"
        self.fresh = True
        self.has_dot = False

    def apply(self, nxt: str) -> None:
        if self.fresh and self.op != "":
            self.op = nxt
            return
        cur = strings.to_float(self.display, 0.0)
        if self.op == "":
            self.acc = cur
        if self.op == "+":
            self.acc = self.acc + cur
        if self.op == "-":
            self.acc = self.acc - cur
        if self.op == "×":
            self.acc = self.acc * cur
        if self.op == "÷":
            if cur == 0.0:
                self.display = "Error"
                self.acc = 0.0
                self.op = ""
                self.fresh = True
                return
            self.acc = self.acc / cur
        self.display = f"{self.acc}"
        self.op = nxt
        self.fresh = True

    def do_op(self, o: str) -> None:
        Calc.apply(o)

    def equals(self) -> None:
        Calc.apply("")

    def clear(self) -> None:
        self.display = "0"
        self.acc = 0.0
        self.op = ""
        self.fresh = True
        self.has_dot = False


key = style(
    size=20, background="panel",
    hover_background="#45475a", active_background="#585b70",
)
fun_tint = style(background="#313244", color="#a6adc8")
fun = key | fun_tint
op_tint = style(
    background="#fab387", color="#1e1e2e",
    hover_background="#f8c49b", active_background="#f5e0dc",
)
op = key | op_tint
readout = style(size=40, color="text", align="right", grow=1.4)


def view():
    with column(spacing=8, padding=16, grow=1):
        text(f"{Calc.display}", **readout)
        with grid(columns=4, rows=5, spacing=8, grow=5):
            button("C", on_click=Calc.clear, **fun)
            button("±", on_click=Calc.negate, **fun)
            button("%", on_click=Calc.percent, **fun)
            button("÷", on_click=lambda: Calc.do_op("÷"), **op)
            button("7", on_click=lambda: Calc.press("7"), **key)
            button("8", on_click=lambda: Calc.press("8"), **key)
            button("9", on_click=lambda: Calc.press("9"), **key)
            button("×", on_click=lambda: Calc.do_op("×"), **op)
            button("4", on_click=lambda: Calc.press("4"), **key)
            button("5", on_click=lambda: Calc.press("5"), **key)
            button("6", on_click=lambda: Calc.press("6"), **key)
            button("-", on_click=lambda: Calc.do_op("-"), **op)
            button("1", on_click=lambda: Calc.press("1"), **key)
            button("2", on_click=lambda: Calc.press("2"), **key)
            button("3", on_click=lambda: Calc.press("3"), **key)
            button("+", on_click=lambda: Calc.do_op("+"), **op)
            button("0", on_click=lambda: Calc.press("0"), col_span=2, **key)
            button(".", on_click=Calc.dot, **key)
            button("=", on_click=Calc.equals, **op)


if __name__ == "__main__":
    run(view, title="calcgrid")

File: gomamochi/demo/calcgrid.go
// The same calculator as demo/calc.go, on a grid instead of five rows.
// `ColSpan` is what makes the zero key twice as wide.
package main

import (
	"strconv"
	"strings"

	. "github.com/i2y/yokan/gomamochi"
)

func key(b *ButtonEl) *ButtonEl {
	return b.Grow(1).Size(20).Background("panel").HoverBackground("#45475a").ActiveBackground("#585b70")
}

func fun(b *ButtonEl) *ButtonEl { return key(b).Background("#313244").Color("#a6adc8") }

func op(b *ButtonEl) *ButtonEl {
	return key(b).Background("#fab387").Color("#1e1e2e").HoverBackground("#f8c49b").ActiveBackground("#f5e0dc")
}

type CalcGrid struct {
	display string
	acc     float64
	op      string
	fresh   bool
	hasDot  bool
}

func num(v float64) string {
	s := strconv.FormatFloat(v, 'f', -1, 64)
	if !strings.ContainsAny(s, ".e") {
		s += ".0"
	}
	return s
}

func toFloatOrZero(s string) float64 {
	if strings.HasSuffix(s, ".") {
		return 0
	}
	v, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0
	}
	return v
}

func (c *CalcGrid) press(d string) {
	if c.fresh {
		c.display = d
		c.fresh = false
		c.hasDot = false
	} else if c.display == "0" {
		c.display = d
	} else {
		c.display = c.display + d
	}
}

func (c *CalcGrid) dot() {
	if c.fresh {
		c.display = "0."
		c.fresh = false
		c.hasDot = true
	} else if !c.hasDot {
		c.display = c.display + "."
		c.hasDot = true
	}
}

func (c *CalcGrid) negate() {
	v := toFloatOrZero(c.display)
	if v == 0 {
		return
	}
	c.display = num(0 - v)
	c.fresh = false
}

func (c *CalcGrid) percent() {
	c.display = num(toFloatOrZero(c.display) / 100)
	c.fresh = true
	c.hasDot = false
}

func (c *CalcGrid) apply(next string) {
	if c.fresh && c.op != "" {
		c.op = next
		return
	}
	cur := toFloatOrZero(c.display)
	switch c.op {
	case "":
		c.acc = cur
	case "+":
		c.acc += cur
	case "-":
		c.acc -= cur
	case "×":
		c.acc *= cur
	case "÷":
		if cur == 0 {
			c.display = "Error"
			c.acc = 0
			c.op = ""
			c.fresh = true
			return
		}
		c.acc /= cur
	}
	c.display = num(c.acc)
	c.op = next
	c.fresh = true
}

func (c *CalcGrid) clear() {
	c.display = "0"
	c.acc = 0
	c.op = ""
	c.fresh = true
	c.hasDot = false
}

func (c *CalcGrid) digit(d string) Element {
	return key(Button(d)).OnClick(func() { c.press(d) })
}

func (c *CalcGrid) opKey(name string) Element {
	return op(Button(name)).OnClick(func() { c.apply(name) })
}

func (c *CalcGrid) View() Element {
	return Column(
		Text(c.display).Size(40).Color("text").Align("right").Grow(1.4),
		Grid(
			fun(Button("C")).OnClick(func() { c.clear() }),
			fun(Button("±")).OnClick(func() { c.negate() }),
			fun(Button("%")).OnClick(func() { c.percent() }),
			c.opKey("÷"),
			c.digit("7"), c.digit("8"), c.digit("9"), c.opKey("×"),
			c.digit("4"), c.digit("5"), c.digit("6"), c.opKey("-"),
			c.digit("1"), c.digit("2"), c.digit("3"), c.opKey("+"),
			key(Button("0")).ColSpan(2).OnClick(func() { c.press("0") }),
			key(Button(".")).OnClick(func() { c.dot() }),
			op(Button("=")).OnClick(func() { c.apply("") }),
		).Columns(4).Rows(5).Spacing(8).Grow(5),
	).Spacing(8).Padding(16).Grow(1)
}

func main() {
	Run(&CalcGrid{display: "0", fresh: true}, Title("calcgrid"))
}
